src/main/scala/nodes/learning/KMeansPlusPlus.scala

```diff
@@ -0,0 +1,149 @@
+package nodes.learning
+
+import breeze.linalg._
+import breeze.stats.distributions.Multinomial
+import breeze.stats.mean
+import org.apache.spark.rdd.RDD
+import workflow.{Estimator, Transformer}
+import utils.MatrixUtils
+
+/**
+ * A KMeans assigning transformer
+ * @param means matrix of dimension numClusters by numFeatures
+ */
+case class KMeansModel(means: DenseMatrix[Double]) extends Transformer[DenseVector[Double], DenseVector[Double]] {
+  /**
+   * Returns the assignment of each vector to the nearest cluster.
+   */
+  def apply(in: DenseVector[Double]): DenseVector[Double] = {
+    // TODO: Could maybe do more efficient single-item implementation
+    apply(in.asDenseMatrix).flatten()
+  }
+
+  def apply(in: DenseMatrix[Double]): DenseMatrix[Double] = {
+    val XSqNormHlf: DenseVector[Double] = sum(in :* in, Axis._1) / 2.0
+    /* compute the distance to all of the centers and assign each point to its nearest center. */
+    val sqDistToCenters = in * means.t
+    sqDistToCenters :*= -1.0
+    sqDistToCenters(::, *) += XSqNormHlf
+    sqDistToCenters(*, ::) += (sum(means :* means, Axis._1) :*= 0.5)
+
+    val nearestCenter = argmin(sqDistToCenters(*, ::))
+
+    // reuse the previous (potentially large) matrix to keep memory usage down
+    val centerAssign = sqDistToCenters
+    var row: Int = 0
+    while (row < in.rows) {
+      var col: Int = 0
+      while (col < means.rows) {
+        centerAssign(row, col) = 0.0
```

Comment:
actually - no, i'm confused by what this is doing - the point of this is to build a binary matrix indicating where all the centers are?